src/astarte-client/client.ts

```diff
@@ -320,17 +329,49 @@ class AstarteClient {
     deviceId: AstarteDevice['id'];
     interfaceName: AstarteInterface['name'];
   }): Promise<AstarteInterfaceValues> {
-    const { deviceId, interfaceName } = params;
     const response = await this.$get(
       this.apiConfig.deviceData({
-        deviceId,
-        interfaceName,
+        deviceId: params.deviceId,
+        interfaceName: params.interfaceName,
         ...this.config,
       }),
     );
     return response.data;
   }
 
+  async getDeviceDataTree(
+    params: DeviceOrDeviceIdParams & InterfaceOrInterfaceNameParams,
+  ): Promise<AstarteDataTreeNode> {
+    let device: AstarteDevice;
+    let iface: AstarteInterface;
+    if ('device' in params) {
+      device = params.device;
+    } else {
+      device = await this.getDeviceInfo(params.deviceId).catch(() => {
+        throw new Error(`Could not find device with ID ${params.deviceId}`);
+      });
+    }
+    if ('interface' in params) {
+      iface = params.interface;
+    } else {
+      const interfaceIntrospection = device.introspection.get(params.interfaceName);
+      if (!interfaceIntrospection) {
+        throw new Error(`Could not find interface ${params.interfaceName} in device introspection`);
+      }
+      iface = await this.getInterface({
+        interfaceName: params.interfaceName,
+        interfaceMajor: interfaceIntrospection.major,
+      }).catch(() => {
+        throw new Error('Could not retrieve interface properties.');
+      });
+    }
+    const interfaceValues = await this.getDeviceData({
+      deviceId: device.id,
+      interfaceName: iface.name,
+    });
+    return toAstarteDataTree({ device, interface: iface, data: interfaceValues });
```

Comment:
this whole method could be simplified assuming the developers provides you the information you need
```
  async getDeviceDataTree({ deviceId, iface }) {
    const interfaceValues = await this.getDeviceData({
      deviceId,
      interfaceName: iface.name,
    });
    return toAstarteDataTree({ iface, data: interfaceValues });
  }
```

if you want to provide developers with utility functions you can do so but relining on the smaller blocks you already created and without duplicating code

```
  fancyGetDeviceData(deviceId, interfaceName) {
    const device = await this.getDeviceInfo(params.deviceId)
    // note that we don't catch the Error, we let it propagate to avoid error handling duplication

    const interfaceIntrospection = device.introspection.get(params.interfaceName);
    if (!interfaceIntrospection) {
      throw new Error(`Could not find interface ${params.interfaceName} in device introspection`);
    }
    const iface = await this.getInterface({
        interfaceName: params.interfaceName,
        interfaceMajor: interfaceIntrospection.major,
      })
    return getDeviceDataTree({ deviceId: device.Id, iface })
  }
```
less code, more readable, and does the same thing